onnxruntime/contrib_ops/cpu/bert/decoder_masked_multihead_attention.cc

```diff
@@ -0,0 +1,473 @@
+    int max_sequence_length,
+    int head_size,
+    const T* past_key_data,
+    T* present_key_data,
+    ThreadPool* tp,
+    const T* attn_bias_data,
+    bool broadcast_attn_bias_dim_0,
+    bool broadcast_attn_bias_dim_1,
+    const int32_t* cache_indir_data,
+    int beam_width,
+    T* scaled_qk_data) const {
+  float scale = scale_ == 0.0f ? 1.0f / sqrt(static_cast<float>(head_size)) : scale_;
+
+  TensorOpCost unit_cost;
+  auto total_sequence_length = past_sequence_length + 1;
+  const ptrdiff_t probs_matrix_size = total_sequence_length;
+  const ptrdiff_t probs_matrix_bytes = probs_matrix_size * sizeof(T);
+
+  unit_cost.compute_cycles = static_cast<double>((SafeInt<ptrdiff_t>(2) * head_size - 1) * total_sequence_length);
+  unit_cost.bytes_loaded = static_cast<double>(SafeInt<ptrdiff_t>(2) * head_size * total_sequence_length * sizeof(T));
+  unit_cost.bytes_stored = static_cast<double>(SafeInt<ptrdiff_t>(head_size) * total_sequence_length * sizeof(T));
+
+  if (attn_bias_data != nullptr) {
+    unit_cost.bytes_loaded += static_cast<double>(probs_matrix_bytes) * 2;
+    unit_cost.bytes_stored += probs_matrix_bytes;
+  }
+
+  if (mask_index_data != nullptr) {
+    unit_cost.bytes_stored += probs_matrix_bytes;
+  }
+
+  // Cost of appending current key to present key
+  unit_cost.compute_cycles += static_cast<double>(head_size);
+  unit_cost.bytes_loaded += static_cast<double>(head_size);
+
+  // Parallel for loop
+  const int loop_len = batch_size * num_heads_;
+  ThreadPool::TryParallelFor(tp, loop_len, unit_cost, [&](std::ptrdiff_t begin, std::ptrdiff_t end) {
+    for (std::ptrdiff_t i = begin; i != end; ++i) {
+      const int batch_index = static_cast<int>(i) / num_heads_;
+      const int head_index = static_cast<int>(i) % num_heads_;
+      const int beam_batch_index = batch_index / beam_width;
+      const T* q_vec = Q + i * head_size;
+      const int attn_bias_base_offset = ((broadcast_attn_bias_dim_0 ? 0 : (beam_batch_index * num_heads_)) +
+                                         (broadcast_attn_bias_dim_1 ? 0 : head_index)) *
+                                        static_cast<int>(probs_matrix_size);
+
+      {
+        // Calculate the latest position of the attention_probs
+        // (1, H) x (T, H)^T -> (1, T)
+        // Decompose into T (1, H) x (1, H)^T -> (1, 1) operations
+        auto last_offset = past_sequence_length + i * static_cast<int>(probs_matrix_size);
+        T* attention_probs_ptr = reinterpret_cast<T*>(attention_probs) + last_offset;
+        math::Dot<float, CPUMathUtil>(head_size, q_vec, K + i * head_size, attention_probs_ptr, nullptr);
+
+        // Apply the attention bias and mask
+        if (attn_bias_data != nullptr) {
+          *attention_probs_ptr += attn_bias_data[attn_bias_base_offset + past_sequence_length];
+        }
+        bool is_masked = (mask_index_data != nullptr) &&
+                         (mask_index_data[(batch_index + 1) * total_sequence_length - 1] == 0);
```

Comment:
The offset `(batch_index + 1) * total_sequence_length - 1` looks not normal. Shall it be `batch_index*total_sequence_length - 1`?